onnx/defs/math/defs.cc

```diff
@@ -3660,4 +3660,534 @@ ONNX_OPERATOR_SET_SCHEMA(
+      return static_cast<T>(ONNX_NAMESPACE::ParseData<float>(t).at(0));
+    case ONNX_NAMESPACE::TensorProto::DOUBLE:
+      return static_cast<T>(ONNX_NAMESPACE::ParseData<double>(t).at(0));
+    case ONNX_NAMESPACE::TensorProto::INT32:
+      return static_cast<T>(ONNX_NAMESPACE::ParseData<int32_t>(t).at(0));
+    case ONNX_NAMESPACE::TensorProto::INT64:
+      return static_cast<T>(ONNX_NAMESPACE::ParseData<int64_t>(t).at(0));
+    default:
+      fail_shape_inference("Unsupported input data type of ", data_type);
+  }
+}
+
+static const char* DFT_ver13_doc =
+    R"DOC(Computes the discrete Fourier transform of input.)DOC";
+
+    ONNX_OPERATOR_SET_SCHEMA(DFT,
+        13,
+        OpSchema()
+            .SetDoc(DFT_ver13_doc)
+            .Attr(
+                "onesided",
+                "If onesided is 1, only values for ω in [0, 1, 2, ..., floor(n_fft/2) + 1] are returned because "
+                "the real-to-complex Fourier transform satisfies the conjugate symmetry, i.e., X[m, ω] = X[m,ω]=X[m,n_fft−ω]*. "
+                "Note if the input or window tensors are complex, then onesided output is not possible. "
+                "Enabling onesided with real inputs performs a Real-valued fast Fourier transform (RFFT)."
+                "When invoked with real or complex valued input, the default value is 0. "
+                "Values can be 0 or 1.",
+                AttributeProto::INT,
+                static_cast<int64_t>(0))
+            .Input(
+                0,
+                "input",
+                "Input tensor representing a real or complex valued signal. "
+                "For real input, the following shape is expected: [batch_size][n_fft]. "
+                "For complex input, the following shape is expected: [batch_size][n_fft][2], where " 
+                "[batch_size][n_fft][0] represents the real component and [batch_size][n_fft][1] represents the imaginary component of the signal.",
+                "T",
+                OpSchema::Single,
+                true,
+                1,
+                OpSchema::NonDifferentiable)
+            .Output(
+                0,
+                "output",
+                "The discrete Fourier transform of input. "
+                "If onesided is 1, the output has the shape: [batch_size][floor(n_fft/2)+1][2]. " 
+                "If onesided is 0, the output has the shape: [batch_size][n_fft][2]"
+                "For all types of input, the last dimension of the output represents the components of a complex number.",
+                "T")
+            .TypeConstraint(
+                "T",
+                {"tensor(float16)", "tensor(float)", "tensor(double)"},
+                "Constrain input and output types to float tensors.")
+            .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+                propagateElemTypeFromInputToOutput(ctx, 0, 0);
+                // To support multidimensional DFT the signal_ndim may != 1
```

Comment:
Is the goal to support multidimensional? If so, my recommendation would be an attribute to indicate if the input is complex or real (and then the number of dimensions is automatically determined by the input rank).

Reply:
Please extend it to multi-dimensional now before adding it to ONNX. Otherwise image processing applications like magnetic resonance image reconstruction (https://fastmri.org/) will have to resort to weird workarounds which will break in upates.

Reply:
I would choose to support multidimensional. We need to choose how to handle complex. Most of these operators return complex. Three options for the return: complex tensor, real tensor + 1 dimension, 2 tensors (real + imaginary parts). Using complex seems the best choice but we need to add new operators (real part, imaginary part, conj). Otherwise we may keep real for now and extend it later to complex.

Reply:
@wuerflts : do you see a need for 2D or 3D, or generic N-D? Do you favor separate ops, or a single op for all dimensions?

Reply:
I think multidim fft can be added if need is there in a separate PR. I am okay either way.